### ISSUE TITLE
DomainResponse should expose a type safe payload

### ISSUE THREAD

**Issue by guilhermeluizsp** [ava-innersource/Liquid-Application-Framework-1.0-deprecated#58]
`DomainResponse` has a `PayLoad` property that represents the actual value of a business operation.
Currently, this property is of type `JToken`, making it awkward to use as we always have to call the [JToken.ToObject](https://www.newtonsoft.com/json/help/html/Overload_Newtonsoft_Json_Linq_JToken_ToObject.htm) method.

https://github.com/Avanade/Liquid-Application-Framework/blob/93f7874f2ead13d3602cc39fc1e9355206217f57/src/Liquid.Domain/Base/Domain/DomainResponse.cs#L25-L28

In my opinion, `DomainResponse` should have two versions: A base one, without the payload, and a second one with a type safe payload property:

```csharp
//extra code removed for readability reasons
public class DomainResponse
{
    public JToken Critics { get; set; }
}

public class DomainResponse<TPayload> : DomainResponse
{
    public TPayload Payload { get; set; }
}
```

### Why

- Not every business operation returns a value. Some of them just indicates if everything went right (e.g "Successfully saved")
- Relying on type safety is always beneficial as we can solve things at compile time, and also makes the code clearer to the developers

### What to consider
- Since this would be a breaking change, maybe we should introduce a new class, considering that `DomainResponse` already has a couple of issues (#57, #54, #35)

**Comment by bruno-brant** [comment]
There's an additional issue in calling `ToObject` all the time: performance.

While I think it's beneficial to have a standardized object to represent the result of a domain computation, I think a great deal of refactoring (or rethinking) can be applied to this. The goal is to simplify communication between a `LightController` (or `LightWorker`) and a domain service (...`LightService`, I guess...?) so that code in the controller can be simplified. 

As such, this class should either be:
1. A (_typed_) object that is the result of the method called, at least for Querying methods
2. _Nothing_, that is, a void method that successfully processed
3. Something to represent know, standard errors, such as _Not Found_, _Bad Request_  (remember that the domain should validate input as well as the controllers) 
4. A _generic business error_

I do have an issue with 4, however: should we be returning an error in that case or should we be throwing exceptions?

**Comment by guilhermeluizsp** [comment]
@bruno-brant I have found that developers usually tend to abuse exceptions by using them to flow control. I personally am not a fan of this approach as it makes the code very awkward to deal with, frequently resulting in try-catch statements everywhere.

I love the concept of "Exceptions are for exceptional circumstances". Why throw an exception when it's expected to fail (validation is an excellent case for this)? A [Result type](https://en.wikipedia.org/wiki/Result_type) (like you mentioned) would perfectly fit in this kind of situations. Not only it would improve performance, but it would make clear to the developer when a method can fail or not.

I'm really into functional programming lately and this one of the core principles of this kind of programming languages: You either have a value or an error code. Do you think we could introduce something like this within Liquid?

**Comment by bruno-brant** [comment]
Yes, exception abusing *is* an issue - the stack unwinding process alone will cost precious CPU-time - but I'm not completely sure how to model such returns in a simpler way -  in practice I think people have a hard time writing typed code that can have numerous returns. 

One the other hand, if we take even half a step towards something that would lead to more try/catches within Liquid applications, then I would just go back and rethink the whole thing. I think user code should have something between zero and none try/catches.

------

A result type might be nice - one could, for instance, leverage an `Either<T1, T2>` type, but I'm sure that's more complex than it sounds; We must remember that one of Liquid's goal is to have a very low learning curve.

We should try and workout some real world cases and generalize from it, what do you think?
